pkg/dedup/dedup_linux.go

```diff
@@ -0,0 +1,150 @@
+package dedup
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"sync"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+type dedupFiles struct {
+	lock          sync.Mutex
+	dev           uint64
+	visitedInodes map[uint64]interface{}
+}
+
+func newDedupFiles() *dedupFiles {
+	return &dedupFiles{
+		visitedInodes: make(map[uint64]interface{}),
+	}
+}
+
+func (d *dedupFiles) checkInode(dev, ino uint64) bool {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.dev != dev {
+		return false
+	}
+
+	_, visited := d.visitedInodes[ino]
+	return visited
+}
+
+func (d *dedupFiles) recordInode(dev, ino uint64) (bool, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	// record the device used by the first inode, do not allow
+	// recording inodes from different devices
+	if d.dev == 0 {
+		d.dev = dev
+	} else if d.dev != dev {
+		return false, errors.New("inode from different device")
+	}
+
+	_, visited := d.visitedInodes[ino]
+	d.visitedInodes[ino] = struct{}{}
+	return visited, nil
+}
+
+// RecordFile records that the file is being processed.  Returns true if the file was already visited.
+func (d *dedupFiles) RecordFile(fi fs.FileInfo) (bool, error) {
+	st, ok := fi.Sys().(*syscall.Stat_t)
+	if !ok {
+		return false, fmt.Errorf("unable to get raw syscall.Stat_t data")
+	}
+	return d.recordInode(uint64(st.Dev), st.Ino)
+}
+
+// Dedup deduplicates the file at src path to dst path
+func (d *dedupFiles) Dedup(src, dst string, fiDst fs.FileInfo) (uint64, error) {
+	srcFile, err := os.OpenFile(src, os.O_RDONLY, 0)
+	if err != nil {
+		return 0, fmt.Errorf("failed to open source file: %w", err)
+	}
+	defer srcFile.Close()
+
+	dstFile, err := os.OpenFile(dst, os.O_WRONLY, 0)
+	if err != nil {
+		return 0, fmt.Errorf("failed to open destination file: %w", err)
+	}
+	defer dstFile.Close()
+
+	stSrc, err := srcFile.Stat()
+	if err != nil {
+		return 0, fmt.Errorf("failed to stat source file: %w", err)
+	}
+	sSrc, ok := stSrc.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, fmt.Errorf("unable to get raw syscall.Stat_t data")
+	}
+	sDest, ok := fiDst.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, fmt.Errorf("unable to get raw syscall.Stat_t data")
+	}
+	if sSrc.Ino == sDest.Ino {
```

Comment:
A comment pointing at the code that handles `Dev` matching, explaining that’s why we don’t need to compare it here, would be useful.

Reply:
I’d prefer this to be cleaned up, maybe reverting `recordInode` to fail again if it sees a cross-device candidate, and this code to point at that `recordInode` condition.